### PR TITLE
[Feature] Add ClearMLLoggerHook

### PR DESCRIPTION
**Pull request by akravchuk97**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

Add `ClearMLLoggerHook` for logging experiments into [ClearML](https://github.com/allegroai/clearml)

## Modification

Please briefly describe what modification is made in this PR.

*  Add `ClearMLLoggerHook` and its unit test

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
Yet another logger :slightly_smiling_face: 
```python
hook = ClearMLLoggerHook(init_kwargs={
    'project_name': 'some_proj',
    'task_name': 'some_task',
})
```
## Checklist

**Before PR**:

-  [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by akravchuk97** [comment]
the precommit has changed `__all__` and imports a lot in some `__init__` files, is that ok?

File: mmcv/runner/__init__.py [hunk 1]
```diff
@@ -44,29 +45,86 @@
...
+    'Sequential',
+    'ModuleDict',
+    'ModuleList',
+    'GradientCumulativeOptimizerHook',
+    'GradientCumulativeFp16OptimizerHook',
+    'DefaultRunnerConstructor',
+    'SegmindLoggerHook',
+    'LinearAnnealingMomentumUpdaterHook',
+    'LinearAnnealingLrUpdaterHook',
+    'ClearMLLoggerHook',
```
Comment (zhouzaida) [inline mmcv/runner/__init__.py:129]:
```suggestion
    'ClearMLLoggerHook'
```

**Comment by zhouzaida** [comment]
> the precommit has changed `__all__` and imports a lot in some `__init__` files, is that ok?

Hi @akravchuk97 , thanks for your contribution. The pre-commit will not change the `__all__` if we remove the comma after `ClearMLLoggerHook`.

File: mmcv/runner/hooks/logger/__init__.py [hunk 2]
```diff
@@ -10,7 +11,14 @@
...
+    'LoggerHook',
+    'MlflowLoggerHook',
+    'PaviLoggerHook',
+    'TensorboardLoggerHook',
+    'TextLoggerHook',
+    'WandbLoggerHook',
+    'NeptuneLoggerHook',
+    'DvcliveLoggerHook',
+    'SegmindLoggerHook',
+    'ClearMLLoggerHook',
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/__init__.py:23]:
```suggestion
    'ClearMLLoggerHook'
```

File: mmcv/runner/hooks/__init__.py [hunk 3]
```diff
@@ -29,19 +30,48 @@
...
+    'EMAHook',
+    'EvalHook',
+    'DistEvalHook',
+    'ProfilerHook',
+    'GradientCumulativeOptimizerHook',
+    'GradientCumulativeFp16OptimizerHook',
+    'SegmindLoggerHook',
+    'LinearAnnealingLrUpdaterHook',
+    'LinearAnnealingMomentumUpdaterHook',
+    'ClearMLLoggerHook',
```
Comment (zhouzaida) [inline mmcv/runner/hooks/__init__.py:76]:
```suggestion
    'ClearMLLoggerHook'
```

File: mmcv/runner/hooks/logger/clearml.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+
+    Args:
+        task_init_kwargs (dict): A dict contains the `clearml.Task.init`
+        initialization keys. Check
+            https://clear.ml/docs/latest/docs/references/sdk/task/#taskinit .
+        interval (int): Logging interval (every k iterations). Default 10.
+        ignore_last (bool): Ignore the log of last iterations in each epoch
+            if less than `interval`.
+            Default: True.
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/clearml.py:22]:
```suggestion
        ignore_last (bool): Ignore the log of last iterations in each epoch
            if less than `interval`. Default: True.
```

File: mmcv/runner/hooks/logger/clearml.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+class ClearMLLoggerHook(LoggerHook):
+    """Class to log metrics with clearml.
+
+    It requires `clearml`_ to be installed.
+
+
+    Args:
+        task_init_kwargs (dict): A dict contains the `clearml.Task.init`
+        initialization keys. Check
+            https://clear.ml/docs/latest/docs/references/sdk/task/#taskinit .
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/clearml.py:18]:
```suggestion
        task_init_kwargs (dict): A dict contains the `clearml.Task.init`
            initialization keys. See `taskinit`_  for more details.
```

File: mmcv/runner/hooks/logger/clearml.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        ignore_last (bool): Ignore the log of last iterations in each epoch
+            if less than `interval`.
+            Default: True.
+        reset_flag (bool): Whether to clear the output buffer after logging.
+            Default: False.
+        by_epoch (bool): Whether EpochBasedRunner is used.
+            Default: True.
+
+    .. _clearml:
+        https://clear.ml/docs/latest/docs/
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/clearml.py:29]:
```suggestion
    .. _clearml:
        https://clear.ml/docs/latest/docs/
    .. _taskinit:
        https://clear.ml/docs/latest/docs/references/sdk/task/#taskinit
```

File: mmcv/runner/hooks/logger/clearml.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+@HOOKS.register_module()
+class ClearMLLoggerHook(LoggerHook):
+    """Class to log metrics with clearml.
+
+    It requires `clearml`_ to be installed.
+
+
+    Args:
+        task_init_kwargs (dict): A dict contains the `clearml.Task.init`
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/clearml.py:16]: Can we rename `task_init_kwargs` to `init_kwargs` which will keep consistent with other logger hooks?

File: mmcv/runner/hooks/logger/clearml.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            raise ImportError(
+                'Please run "pip install clearml" to install clearml')
+        self.clearml = clearml
+
+    @master_only
+    def before_run(self, runner):
+        super(ClearMLLoggerHook, self).before_run(runner)
+        task_kwargs = self.task_init_kwargs if self.task_init_kwargs else {}
+        self.task = self.clearml.Task.init(**task_kwargs)
+        self.logger = self.task.get_logger()
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/clearml.py:56]:
```suggestion
        self.task_logger = self.task.get_logger()
```

File: mmcv/runner/__init__.py (hunk 1 above)
Comment (akravchuk97) [reply mmcv/runner/__init__.py]: done

File: mmcv/runner/hooks/logger/__init__.py (hunk 2 above)
Comment (akravchuk97) [reply mmcv/runner/hooks/logger/__init__.py]: done

File: mmcv/runner/hooks/__init__.py (hunk 3 above)
Comment (akravchuk97) [reply mmcv/runner/hooks/__init__.py]: done

File: mmcv/runner/hooks/logger/clearml.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        initialization keys. Check
+            https://clear.ml/docs/latest/docs/references/sdk/task/#taskinit .
+        interval (int): Logging interval (every k iterations). Default 10.
+        ignore_last (bool): Ignore the log of last iterations in each epoch
+            if less than `interval`.
+            Default: True.
+        reset_flag (bool): Whether to clear the output buffer after logging.
+            Default: False.
+        by_epoch (bool): Whether EpochBasedRunner is used.
+            Default: True.
```
[the remark this answers is not in the document]
Comment (akravchuk97) [reply mmcv/runner/hooks/logger/clearml.py]: done

File: mmcv/runner/hooks/logger/clearml.py (hunk 4 above)
Comment (akravchuk97) [reply mmcv/runner/hooks/logger/clearml.py]: done

File: mmcv/runner/hooks/logger/clearml.py (hunk 5 above)
Comment (akravchuk97) [reply mmcv/runner/hooks/logger/clearml.py]: done

File: mmcv/runner/hooks/logger/clearml.py (hunk 6 above)
Comment (akravchuk97) [reply mmcv/runner/hooks/logger/clearml.py:26]: done

File: mmcv/runner/hooks/logger/clearml.py (hunk 8 above)
Comment (akravchuk97) [reply mmcv/runner/hooks/logger/clearml.py]: done

File: mmcv/runner/hooks/logger/clearml.py (hunk 7 above)
Comment (akravchuk97) [reply mmcv/runner/hooks/logger/clearml.py]: done

File: mmcv/runner/hooks/logger/clearml.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+@HOOKS.register_module()
+class ClearMLLoggerHook(LoggerHook):
+    """Class to log metrics with clearml.
+
+    It requires `clearml`_ to be installed.
+
+
+    Args:
+        init_kwargs (dict): A dict contains the `clearml.Task.init`
+        initialization keys. See `taskinit`_  for more details.
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/clearml.py:17]:
```suggestion
        init_kwargs (dict): A dict contains the `clearml.Task.init`
            initialization keys. See `taskinit`_  for more details.
```

Comment (akravchuk97) [reply mmcv/runner/hooks/logger/clearml.py]: done

**Comment by zhouzaida** [comment]
Hi @akravchuk97 , is it convenient to provide a shareable experiment log here so that users can see the results generated by ClearMLLoggerHook.

**Comment by akravchuk97** [comment]
Hi @zhouzaida !
Will there be ok such an example? 
[Link](https://app.clear.ml/projects/3bc120a534f04712b7abdf2b36ad1739/experiments/2ec9a41d79fc4f6c8268f95501d9d53a/output/execution) to the log for this [tutorial](https://github.com/open-mmlab/mmdetection/blob/master/demo/MMDet_InstanceSeg_Tutorial.ipynb) from mmdetection.
To open it, you need to log in to clearml. Then open Results->Scalars to  see the graphs.

**Comment by zhouzaida** [comment]
> Hi @zhouzaida ! Will there be ok such an example? [Link](https://app.clear.ml/projects/3bc120a534f04712b7abdf2b36ad1739/experiments/2ec9a41d79fc4f6c8268f95501d9d53a/output/execution) to the log for this [tutorial](https://github.com/open-mmlab/mmdetection/blob/master/demo/MMDet_InstanceSeg_Tutorial.ipynb) from mmdetection. To open it, you need to log in to clearml. Then open Scalars to see the graphs.

Yet. The above example is enough. Thanks for your contribution again.

**Review by zhouzaida — APPROVED** [review]

**Review by teamwong111 — APPROVED** [review]

**Review by imabackstabber — APPROVED** [review]
LGTM.

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @arkravchuk ！First of all, we want to express our gratitude for your significant PR in the mmcv project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**